pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -218,21 +218,30 @@ async def _prepare_request_parameters(
     ctx: GraphRunContext[GraphAgentState, GraphAgentDeps[DepsT, NodeRunEndT]],
 ) -> models.ModelRequestParameters:
     """Build tools and create an agent model."""
-    function_tool_defs: list[ToolDefinition] = []
+    function_tool_defs: dict[str, ToolDefinition] = {}
 
     run_context = build_run_context(ctx)
 
     async def add_tool(tool: Tool[DepsT]) -> None:
         ctx = run_context.replace_with(retry=tool.current_retry, tool_name=tool.name)
         if tool_def := await tool.prepare_tool_def(ctx):
-            function_tool_defs.append(tool_def)
+            # prepare_tool_def may change tool_def.name
+            if tool_def.name in function_tool_defs:
+                raise exceptions.UserError(
+                    f"Tool name '{tool_def.name}' for tool `{tool}` conflicts with existing tool: {tool_def.name!r}."
```

Comment:
@Wh1isper Would this work?

```suggestion
                    f"New tool name '{tool_def.name}' for tool `{tool.name}` conflicts with existing tool."
```